lib/Sema/TypeCheckAccess.cpp

```diff
@@ -1421,43 +1426,104 @@ class ImplementationOnlyImportChecker
+        if (!SF.isImportedImplementationOnly(M))
           return true;
 
-        diagnose(component->getBoundDecl(), component);
+        diagnoseType(component->getBoundDecl(), component);
         foundAnyIssues = true;
         // We still continue even in the diagnostic case to report multiple
         // violations.
         return true;
       }));
     }
 
-    // FIXME: Are there places where we can be sure the TypeRepr fully accounts
-    // for the type, and therefore we could skip this extra work?
-    if (!foundAnyIssues && type) {
-      type.walk(SimpleTypeDeclFinder([&](const TypeDecl *typeDecl) {
+    // Note that if we have a type, we can't skip checking it even if the
+    // TypeRepr is okay, because that's how we check what conformances are
+    // being used.
+    //
+    // We still don't want to do this if we found issues with the TypeRepr,
+    // though, because that would result in some issues being reported twice.
+    if (foundAnyIssues || !type)
+      return;
+
+    class ProblematicTypeFinder : public TypeDeclFinder {
+      const SourceFile &SF;
+      CheckImplementationOnlyTypeCallback diagnoseType;
+      CheckImplementationOnlyConformanceCallback diagnoseConformance;
+    public:
+      ProblematicTypeFinder(
+          const SourceFile &SF,
+          CheckImplementationOnlyTypeCallback diagnoseType,
+          CheckImplementationOnlyConformanceCallback diagnoseConformance)
+        : SF(SF), diagnoseType(diagnoseType),
+          diagnoseConformance(diagnoseConformance) {}
+
+      void visitTypeDecl(const TypeDecl *typeDecl) {
         ModuleDecl *M = typeDecl->getModuleContext();
-        if (!SF->isImportedImplementationOnly(M))
-          return TypeWalker::Action::Continue;
+        if (!SF.isImportedImplementationOnly(M))
+          return;
 
-        diagnose(typeDecl, /*typeRepr*/nullptr);
-        // We still continue even in the diagnostic case to report multiple
-        // violations.
-        return TypeWalker::Action::Continue;
-      }));
-    }
+        diagnoseType(typeDecl, /*typeRepr*/nullptr);
+      }
+
+      void visitSubstitutionMap(const SubstitutionMap &subs) {
+        for (ProtocolConformanceRef conformance : subs.getConformances()) {
+          if (!conformance.isConcrete())
+            continue;
+          const ProtocolConformance *concreteConf = conformance.getConcrete();
+
+          SubstitutionMap subConformanceSubs =
+              concreteConf->getSubstitutions(SF.getParentModule());
+          visitSubstitutionMap(subConformanceSubs);
+
+          const RootProtocolConformance *rootConf =
+              concreteConf->getRootConformance();
+          ModuleDecl *M = rootConf->getDeclContext()->getParentModule();
+          if (!SF.isImportedImplementationOnly(M))
+            continue;
+          diagnoseConformance(rootConf);
+        }
+      }
+
+      Action visitNominalType(NominalType *ty) override {
+        visitTypeDecl(ty->getDecl());
```

Comment:
@AnthonyLatsis's PR adds where clauses for types without generic parameter lists, so you should just duplicate the bound generic type logic here; the nominal type might add a conformance that's not there on the parent type.

We could also skip visiting the parent type as an optimization but I don't know how.

(I'd really like to merge NominalType/BoundGenericType into a single type, remove the parent type concept altogether, and store the SubstitutionMap directly inside the type...)

Reply:
Uh, hm. Nominal types don't carry that information, so unless the current DeclContext is now relevant I don't see how the current logic will work with that. Are you sure that's something that can come up in this part of the code? (as opposed to being checked at the FuncDecl or whatever when we walk the requirements)